scarlet/renderer.py

```diff
@@ -0,0 +1,520 @@
+        else:
+            img = image[band]
+        apply_filter(
+            img,
+            psf[band].reshape(-1),
+            bounds[0],
+            bounds[1],
+            bounds[2],
+            bounds[3],
+            result[band],
+        )
+    return result
+
+
+def _grad_convolve(convolved, image, psf, slices):
+    """Gradient of a real space convolution
+    """
+    return lambda input_grad: convolve(input_grad, psf[:, ::-1, ::-1], slices)
+
+
+# Register this function in autograd
+defvjp(convolve, _grad_convolve)
+
+# match the spatial shapes of model and data
+@primitive
+def match_shape(model, data_frame, slices):
+    data_slices, model_slices = slices
+    data_shape = data_frame.shape
+
+    # check if data get sliced
+    if any(
+        [
+            data_slices[d].stop - data_slices[d].start != data_shape[d]
+            for d in range(-2, 0)
+        ]
+    ):
+        matched = np.zeros(data_frame.shape, dtype=data_frame.dtype)
+        matched[data_slices] = model[model_slices]
+        return matched
+
+    return model[model_slices]
+
+
+def _grad_match_shape(upstream_grad, model, data_frame, slices):
+    # just slices gradients like the model
+    data_slices, model_slices = slices
+
+    def result(upstream_grad):
+        _result = np.zeros(model.shape, dtype=model.dtype)
+        _result[model_slices] = upstream_grad[data_slices]
+        return _result
+
+    return result
+
+
+defvjp(match_shape, _grad_match_shape)
+
+
+class ConvolutionRenderer(Renderer):
+    def __init__(self, data_frame, model_frame, convolution_type="fft", padding=10):
+        """
+        convolution_type: str
+            The type of convolution to use.
+            - `real`: Use a real space convolution and gradient
+            - `fft`: Use a DFT to do the convolution and gradient
+        """
+        super().__init__(data_frame, model_frame)
+
+        assert convolution_type in [
+            "real",
+            "fft",
+        ], "`convolution` must be either 'real' or 'fft'"
+        self._convolution_type = convolution_type
+
+        # 2D spatial region covered by data
+        pixel_in_model_frame = data_frame.convert_pixel_to(model_frame)
+        # since there cannot be rotation or scaling, it's only translation
+        ll = np.round(pixel_in_model_frame.min(axis=0)).astype("int")
+        ur = np.round(pixel_in_model_frame.max(axis=0)).astype("int") + 1
+        bounds = (ll[0], ur[0]), (ll[1], ur[1])
+        data_box = model_frame.bbox[0] @ Box.from_bounds(*bounds)
+        # properly treats truncation in both boxes
+        self.slices = overlapped_slices(data_box, model_frame.bbox)
+
+        # construct diff kernel
+        psf_fft = fft.Fourier(data_frame.psf.get_model().astype(model_frame.dtype))
+        model_psf_fft = fft.Fourier(
+            model_frame.psf.get_model().astype(model_frame.dtype)
+        )
+        self.diff_kernel = fft.match_psf(psf_fft, model_psf_fft, padding=padding)
```

Comment:
Shouldn't this have an underscore, since we don't expect the user to update this without catastrophic failure?

Reply:
I don't think so. If a user wanted to overwrite the diff kernel, would the underscore stop them? Since this is moved from Observation to Renderer, I think it is reasonable to expose it.